std/range/primitives.d

```diff
@@ -1664,6 +1674,154 @@ if (isInputRange!Range)
 }
 
 /**
+This is a best-effort implementation of `back` for any kind of
+range. It expects a non-empty range.
+
+If `isBidirectional!Range`, simply returns `range.back` without
+checking `upTo) (when specified).
+
+Otherwise, walks the range through its length and returns the last element seen.
+Performes $(BIGOH n) evaluations of `range.empty` and `range.popFront()`,
+where `n` is the effective length of `range`.
+
+The `upTo` parameter is useful to "cut the losses" in case
+the interest is in seeing whether the range has at least some number
+of elements. If the parameter `upTo` is specified, stops if `upTo`
+steps have been taken and returns `upTo`.
+
+Infinite ranges are compatible, provided the parameter `upTo` is
+specified, in which case the implementation simply returns `upTo`.
+
+Params:
+    range = range to walk through
+
+Returns: last element of the range
+
+See_Also: $(REF walkLength, std, range, primitives)
+*/
+auto walkBack(Range)(Range range)
+if ((isInputRange!Range || isIterable!Range) && !isInfinite!Range)
+in
+{
+
+    assert(!range.empty, "Can't walkBack an empty range");
+}
+body
+{
+    static if (isBidirectionalRange!Range)
+    {
+        return range.back;
+    }
+    else
+    {
+        // Rebindable doesn't support all types as of now
+        Unqual!(ElementType!Range) last;
+        foreach (e; range)
+            last = cast(Unqual!(ElementType!Range)) e;
```

Comment:
I'm having issues testing using your code. As it was, the first `while` test was consuming the range and subsequent tests failed because the range had been consumed.

I fixed it by preceding each call with `auto r = range;` and operating on `r` instead, but this served only to make the `while` approach dramatically slower while not significantly affecting the performance of the `foreach` test.

I would've expected that rewriting like so might make their performance more similar, but I'm surprised the compiler wasn't able to optimize the structure to look more like this anyway:

``` D
bool empty = range.empty;
while(!empty){
    auto element = range.front;
    range.popFront();
    empty = range.empty;
    if(empty) return element;
}
```

@UplinkCoder can you explain what's going on here?